src/plugins/ipt/ipt.cpp

```diff
@@ -0,0 +1,325 @@
+    return (static_cast<uint64_t>(cmd) << 32) | data;
+}
+
+void emit_ptwrite64(std::ofstream& stream, uint64_t payload)
+{
+    // ptwrite packet
+    stream.put(0x02);
+    // no FUP, 8 bytes of payload
+    stream.put(0x32);
+    stream.write(reinterpret_cast<char*>(&payload), sizeof(payload));
+}
+
+
+event_response_t ipt_cr3_cb(drakvuf_t drakvuf, drakvuf_trap_info_t* info)
+{
+    auto plugin = get_trap_plugin<ipt>(info);
+    auto& vcpu = plugin->vcpus[info->vcpu];
+
+    vcpu.flush(info->regs->vmtrace_pos);
+
+    vcpu.annotate(pack_payload(PTW_CURRENT_CR3, info->regs->cr3));
+    vcpu.annotate(pack_payload(PTW_CURRENT_TID, info->proc_data.tid));
+
+    return VMI_EVENT_RESPONSE_NONE;
+}
+
+event_response_t ipt_catchall_cb(drakvuf_t drakvuf, drakvuf_trap_info_t* info)
+{
+    auto plugin = get_trap_plugin<ipt>(info);
+    auto& vcpu = plugin->vcpus[info->vcpu];
+
+    vcpu.flush(info->regs->vmtrace_pos);
+
+    vcpu.annotate(pack_payload(PTW_EVENT_ID, info->event_uid));
+
+    return VMI_EVENT_RESPONSE_NONE;
+}
+
+} // unnamed namespace
+
+void ipt_vcpu::annotate(uint64_t payload)
+{
+    emit_ptwrite64(this->output_stream, payload);
+}
+
+void ipt_vcpu::flush(uint64_t offset)
+{
+    // update last offset
+    uint64_t prev = this->last_offset;
+    this->last_offset = offset;
+
+    PRINT_DEBUG("[IPT] Flushing vCPU %d offset: %" PRIx64 " last offset: %" PRIx64 "\n",
+                id, offset, prev);
+
+    if (!this->output_stream.good())
+    {
+        throw -1;
+    }
+
+    // Cast uint8_t* to char* to satisfy std::ofstream requirements
+    // https://stackoverflow.com/questions/16260033/reinterpret-cast-between-char-and-stduint8-t-safe
+    auto data = reinterpret_cast<char*>(this->buf);
+    if (offset > prev)
+    {
+        // Normal case, some data was appended to buffer
+        this->output_stream.write(data + prev, offset - prev);
+    }
+    else if (offset < prev)
+    {
+        // Buffer wrapped - write from last offset to the end of the buffer
+        // and then from the beginning to last written packet
+        // This assumes that IPT buffer is large enough to not overflow between
+        // calls to ipt_annotate
+        this->output_stream.write(data + prev, this->size - prev);
+        this->output_stream.write(data, offset);
+    }
+    else
+    {
+        PRINT_DEBUG("[IPT] flush_ipt_stream() called but no new IPT data is present\n");
+        // In theory this should be unreachable, since it's hard to generate
+        // no data between events. Handle this, just in case something is wrong
+        this->annotate(pack_payload(PTW_ERROR_EMPTY, 0));
+    }
+}
+
+
+drakvuf_trap_t* reg_cr3_trap(drakvuf_t drakvuf, drakvuf_trap_info_t* info, drakvuf_trap_t* trap)
+{
+    trap->type = REGISTER;
+    trap->reg = CR3;
+
+    if (!drakvuf_add_trap(drakvuf, trap))
+        return nullptr;
+
+    return trap;
+}
+
+drakvuf_trap_t* reg_catchall_trap(drakvuf_t drakvuf, drakvuf_trap_info_t* info, drakvuf_trap_t* trap)
```

Comment:
```suggestion
drakvuf_trap_t* ipt::reg_catchall_trap(drakvuf_t drakvuf, drakvuf_trap_info_t* info, drakvuf_trap_t* trap)
```
As I requested please make this a class static function.

Reply:
done :+1: